### PR TITLE
[RFC] target/arc: Fix off-by-one error in arc_save_context()

### PR DESCRIPTION
**Pull request by artemiy-volkov**
While not affecting the function's main purpose, an error has crept into arc_save_context() that results in logging wrong register values when the debug level is 3 or more. For instance, when debugging a trivial program and halting at entry to main, the following values are printed to the log:

Debug: 2868 5634 arc.c:955 arc_save_context(): Get core register regnum=0, name=r0, value=0x00000000
...
Debug: 2900 5635 arc.c:955 arc_save_context(): Get core register regnum=60, name=lp_count, value=0x900002c4
Debug: 2901 5635 arc.c:955 arc_save_context(): Get core register regnum=63, name=pcl, value=0xffffffff
Debug: 2902 5635 arc.c:970 arc_save_context(): Get aux register regnum=64, name=pc, value=0x900000b4
Debug: 2903 5635 arc.c:970 arc_save_context(): Get aux register regnum=65, name=lp_start, value=0x900000bc
Debug: 2904 5635 arc.c:970 arc_save_context(): Get aux register regnum=66, name=lp_end, value=0x00080801
Debug: 2905 5635 arc.c:970 arc_save_context(): Get aux register regnum=67, name=status32, value=0xffffffff

The off-by-one error is clearly visible. After the change, the register contents make much more sense:

Debug: 2832 3163 arc.c:955 arc_save_context(): Get core register regnum=0, name=r0, value=0x00000000
...
Debug: 2864 3163 arc.c:955 arc_save_context(): Get core register regnum=60, name=lp_count, value=0x00000000
Debug: 2865 3163 arc.c:955 arc_save_context(): Get core register regnum=63, name=pcl, value=0x900002c4
Debug: 2866 3163 arc.c:970 arc_save_context(): Get aux register regnum=64, name=pc, value=0x900002c6
Debug: 2867 3163 arc.c:970 arc_save_context(): Get aux register regnum=65, name=lp_start, value=0x900000b4
Debug: 2868 3163 arc.c:970 arc_save_context(): Get aux register regnum=66, name=lp_end, value=0x900000bc
Debug: 2869 3163 arc.c:970 arc_save_context(): Get aux register regnum=67, name=status32, value=0x00080801

While at it, simplify the function by e.g. removing a superfluous loop counter.

Change-Id: I8f2d79404707fbac4503af45b393ea73f91e6beb

### CODE REVIEW

**Review by abrodkin — APPROVED** [review]
LGTM

**Review by claziss — APPROVED** [review]

**Review by EvgeniiDidin — APPROVED** [review]

**Review by shahab-vahedi — COMMENTED** [review]
Looks good to me. I have two minor remarks:

1) It not necessarily off-by-**ONE**. It can be be off by **many**.
2) Looking at the first instance of printed registers, it is not *clearly* visible that there is off-by-one to the naked eyes. I would rephrase it as "This is the bad output", whereas it should have been ...

**Comment by artemiy-volkov** [comment]
1. I'm afraid it really is off-by-one here. Here's a more contrived example (stop just inside factorial() from the hbreak2.exp test case):

before the change
Debug: 2895 2426 arc.c:894 arc_save_context(): Get core register regnum=0, name=r0, value=0x9000813c
Debug: 2896 2426 arc.c:894 arc_save_context(): Get core register regnum=1, name=r1, value=0x00000006
Debug: 2897 2426 arc.c:894 arc_save_context(): Get core register regnum=2, name=r2, value=0x0000000a
Debug: 2898 2426 arc.c:894 arc_save_context(): Get core register regnum=3, name=r3, value=0x00000001
Debug: 2899 2426 arc.c:894 arc_save_context(): Get core register regnum=4, name=r4, value=0x0ccccccc
Debug: 2900 2426 arc.c:894 arc_save_context(): Get core register regnum=5, name=r5, value=0x0000000a
Debug: 2901 2426 arc.c:894 arc_save_context(): Get core register regnum=6, name=r6, value=0xffffffff
Debug: 2902 2426 arc.c:894 arc_save_context(): Get core register regnum=7, name=r7, value=0x7fffffff
Debug: 2903 2426 arc.c:894 arc_save_context(): Get core register regnum=8, name=r8, value=0x00000007
Debug: 2904 2426 arc.c:894 arc_save_context(): Get core register regnum=9, name=r9, value=0x00000000
Debug: 2905 2426 arc.c:894 arc_save_context(): Get core register regnum=10, name=r10, value=0x00000000
Debug: 2906 2426 arc.c:894 arc_save_context(): Get core register regnum=11, name=r11, value=0xffffff9f
Debug: 2907 2426 arc.c:894 arc_save_context(): Get core register regnum=12, name=r12, value=0x00000000
Debug: 2908 2426 arc.c:894 arc_save_context(): Get core register regnum=13, name=r13, value=0x00000000
Debug: 2909 2426 arc.c:894 arc_save_context(): Get core register regnum=14, name=r14, value=0x00000000
Debug: 2910 2426 arc.c:894 arc_save_context(): Get core register regnum=15, name=r15, value=0x00000000
Debug: 2911 2426 arc.c:894 arc_save_context(): Get core register regnum=16, name=r16, value=0xdfffffe0
Debug: 2912 2426 arc.c:894 arc_save_context(): Get core register regnum=17, name=r17, value=0x00000000
Debug: 2913 2426 arc.c:894 arc_save_context(): Get core register regnum=18, name=r18, value=0x00000000
Debug: 2914 2426 arc.c:894 arc_save_context(): Get core register regnum=19, name=r19, value=0x00000000
Debug: 2915 2426 arc.c:894 arc_save_context(): Get core register regnum=20, name=r20, value=0x00000000
Debug: 2916 2426 arc.c:894 arc_save_context(): Get core register regnum=21, name=r21, value=0x00000000
Debug: 2917 2426 arc.c:894 arc_save_context(): Get core register regnum=22, name=r22, value=0x00000000
Debug: 2918 2426 arc.c:894 arc_save_context(): Get core register regnum=23, name=r23, value=0x00000000
Debug: 2919 2426 arc.c:894 arc_save_context(): Get core register regnum=24, name=r24, value=0x00000000
Debug: 2920 2426 arc.c:894 arc_save_context(): Get core register regnum=25, name=r25, value=0x900090a4
Debug: 2921 2426 arc.c:894 arc_save_context(): Get core register regnum=26, name=gp, value=0xdfffffe0
Debug: 2922 2426 arc.c:894 arc_save_context(): Get core register regnum=27, name=fp, value=0xdfffffb8
Debug: 2923 2426 arc.c:894 arc_save_context(): Get core register regnum=28, name=sp, value=0x00000000
Debug: 2924 2426 arc.c:894 arc_save_context(): Get core register regnum=29, name=ilink, value=0x00000000
Debug: 2925 2426 arc.c:894 arc_save_context(): Get core register regnum=30, name=r30, value=0x90000312
Debug: 2926 2426 arc.c:894 arc_save_context(): Get core register regnum=31, name=blink, value=0x00000000
Debug: 2927 2426 arc.c:894 arc_save_context(): Get core register regnum=60, name=lp_count, value=0x90000368
Debug: 2928 2426 arc.c:894 arc_save_context(): Get core register regnum=63, name=pcl, value=0xffffffff
Debug: 2929 2426 arc.c:909 arc_save_context(): Get aux register regnum=64, name=pc, value=0x900000b4
Debug: 2930 2426 arc.c:909 arc_save_context(): Get aux register regnum=65, name=lp_start, value=0x900000bc
Debug: 2931 2426 arc.c:909 arc_save_context(): Get aux register regnum=66, name=lp_end, value=0x00080801
Debug: 2932 2426 arc.c:909 arc_save_context(): Get aux register regnum=67, name=status32, value=0xffffffff

after the change
Debug: 2895 8274 arc.c:889 arc_save_context(): Get core register regnum=0, name=r0, value=0x00000006
Debug: 2896 8274 arc.c:889 arc_save_context(): Get core register regnum=1, name=r1, value=0x9000813c
Debug: 2897 8274 arc.c:889 arc_save_context(): Get core register regnum=2, name=r2, value=0x00000006
Debug: 2898 8274 arc.c:889 arc_save_context(): Get core register regnum=3, name=r3, value=0x0000000a
Debug: 2899 8274 arc.c:889 arc_save_context(): Get core register regnum=4, name=r4, value=0x00000001
Debug: 2900 8274 arc.c:889 arc_save_context(): Get core register regnum=5, name=r5, value=0x0ccccccc
Debug: 2901 8274 arc.c:889 arc_save_context(): Get core register regnum=6, name=r6, value=0x0000000a
Debug: 2902 8274 arc.c:889 arc_save_context(): Get core register regnum=7, name=r7, value=0xffffffff
Debug: 2903 8274 arc.c:889 arc_save_context(): Get core register regnum=8, name=r8, value=0x7fffffff
Debug: 2904 8274 arc.c:889 arc_save_context(): Get core register regnum=9, name=r9, value=0x00000007
Debug: 2905 8274 arc.c:889 arc_save_context(): Get core register regnum=10, name=r10, value=0x00000000
Debug: 2906 8274 arc.c:889 arc_save_context(): Get core register regnum=11, name=r11, value=0x00000000
Debug: 2907 8274 arc.c:889 arc_save_context(): Get core register regnum=12, name=r12, value=0xffffff9f
Debug: 2908 8274 arc.c:889 arc_save_context(): Get core register regnum=13, name=r13, value=0x00000000
Debug: 2909 8274 arc.c:889 arc_save_context(): Get core register regnum=14, name=r14, value=0x00000000
Debug: 2910 8274 arc.c:889 arc_save_context(): Get core register regnum=15, name=r15, value=0x00000000
Debug: 2911 8274 arc.c:889 arc_save_context(): Get core register regnum=16, name=r16, value=0x00000000
Debug: 2912 8274 arc.c:889 arc_save_context(): Get core register regnum=17, name=r17, value=0xdfffffe0
Debug: 2913 8274 arc.c:889 arc_save_context(): Get core register regnum=18, name=r18, value=0x00000000
Debug: 2914 8274 arc.c:889 arc_save_context(): Get core register regnum=19, name=r19, value=0x00000000
Debug: 2915 8274 arc.c:889 arc_save_context(): Get core register regnum=20, name=r20, value=0x00000000
Debug: 2916 8274 arc.c:889 arc_save_context(): Get core register regnum=21, name=r21, value=0x00000000
Debug: 2917 8274 arc.c:889 arc_save_context(): Get core register regnum=22, name=r22, value=0x00000000
Debug: 2918 8274 arc.c:889 arc_save_context(): Get core register regnum=23, name=r23, value=0x00000000
Debug: 2919 8274 arc.c:889 arc_save_context(): Get core register regnum=24, name=r24, value=0x00000000
Debug: 2920 8274 arc.c:889 arc_save_context(): Get core register regnum=25, name=r25, value=0x00000000
Debug: 2921 8274 arc.c:889 arc_save_context(): Get core register regnum=26, name=gp, value=0x900090a4
Debug: 2922 8274 arc.c:889 arc_save_context(): Get core register regnum=27, name=fp, value=0xdfffffe0
Debug: 2923 8274 arc.c:889 arc_save_context(): Get core register regnum=28, name=sp, value=0xdfffffb8
Debug: 2924 8274 arc.c:889 arc_save_context(): Get core register regnum=29, name=ilink, value=0x00000000
Debug: 2925 8274 arc.c:889 arc_save_context(): Get core register regnum=30, name=r30, value=0x00000000
Debug: 2926 8274 arc.c:889 arc_save_context(): Get core register regnum=31, name=blink, value=0x90000312
Debug: 2927 8274 arc.c:889 arc_save_context(): Get core register regnum=60, name=lp_count, value=0xffffffff
Debug: 2928 8274 arc.c:889 arc_save_context(): Get core register regnum=63, name=pcl, value=0xffffffff
Debug: 2929 8274 arc.c:902 arc_save_context(): Get aux register regnum=64, name=pc, value=0x9000036a
Debug: 2930 8274 arc.c:902 arc_save_context(): Get aux register regnum=65, name=lp_start, value=0x900000b4
Debug: 2931 8274 arc.c:902 arc_save_context(): Get aux register regnum=66, name=lp_end, value=0x900000bc
Debug: 2932 8274 arc.c:902 arc_save_context(): Get aux register regnum=67, name=status32, value=0x00080801

2. I will amend the statement about the error being clearly visible, thank you for the recommendation.
